File: solutions/152. Maximum Product Subarray.cpp
#define fo(i, k, n) for(int i = k; i < n; i++)
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int ans = nums[0], rmx = nums[0], rmn = nums[0];
        fo(i, 1, n){
            int cmax = rmx;
            rmx = max({rmx*nums[i], nums[i], rmn*nums[i]});
            rmn = min({cmax*nums[i], rmn*nums[i], nums[i]});
            ans = max(rmx,ans);
        }
        return ans ;
    }
};

File: solutions/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        // declare vars.
        int gmax=Integer.MIN_VALUE;
        int prefixprod=1;
        int suffixprod=1;
        
        // iterate on the nums
        for (int i=0; i <nums.length; i++) {
            int elem=nums[i];
            if (prefixprod==0)prefixprod=1;
            if (suffixprod==0)suffixprod=1;
            
            // update the prefix and suffix
            prefixprod = prefixprod * elem;
            suffixprod = suffixprod * nums[nums.length - i - 1];
            
            // update gmax
            gmax=Math.max(gmax, Math.max(prefixprod, suffixprod));
        }
        
        // return gmax
        return gmax;
    }
}
